### ISSUE TITLE
404 detection broken, causing application crashes (500 Internal Server Error)

### ISSUE THREAD

**Issue by stomar** [adaoraul/rack-jekyll#35]
The detection of missing files is broken. Every path that is only part of an existing path is assumed to be valid, which leads to application crashes because the corresponding file can not be found.

```
 $ curl -I https://www.ruby-lang.org/en/downloads
     # -> 302 Found: ok (existing path)

 $ curl -I https://www.ruby-lang.org/down
     # -> 500 Internal Server Error: WRONG (should be 404)
     #    non-existing path that is part of an existing one

 $ curl -I https://www.ruby-lang.org/dOWn
     # -> 404 Not Found: ok (non-existing path)
```

The bug is caused by lib/rack/jekyll.rb lines 67 and following:

``` ruby
  @files = ::Dir[@path + "/**/*"].inspect if @files == "[]"
  if @files.include?(path_info)
    # tries to open file -> Errno::ENOENT: No such file or directory @ rb_sysopen
  else
    # handles 404's
  end
```

This checks whether the request path is **part of a string that contains all filenames** in the _site dir!
The test doesn't work properly and assumes that paths are valid in many cases where it shouldn't.

**Comment by greyshi** [comment]
Is there a way to patch this bug manually on Heroku until the fix is on the main branch?

**Comment by stomar** [comment]
You mean generally? Sure:
fork the repo and create a branch that fixes the bug; then specify your repo and branch in the Gemfile,
see https://github.com/ruby/www.ruby-lang.org/blob/master/Gemfile#L8 for an example (my fork fixes a different problem of rack-jekyll, though, and might be removed anytime we find a better solution, so don't use it).

**Comment by stomar** [comment]
PS. It seems maintenance of rack-jekyll is rather dead, so I don't expect a quick fix.

**Comment by stomar** [comment]
duplicate of #23

**Comment by stomar** [comment]
Update: This is now WIP targeted for the next release.

**Comment by stomar** [comment]
Fixed with a9eb164d6c1ec548aa1ce3fc18e6f9f209d01487.
